qi-lib/flow/core/deforest-cps.rkt

```diff
@@ -0,0 +1,323 @@
+#lang racket/base
+
+(provide (for-syntax deforest-pass))
+
+(require (for-syntax racket/base
+                     syntax/parse
+                     "deforest-syntax.rkt"
+                     "../extended/util.rkt"
+                     syntax/srcloc
+                     racket/syntax-srcloc)
+         "deforest-templates.rkt"
+         racket/performance-hint
+         racket/match
+         racket/contract/base)
+
+;; "Composes" higher-order functions inline by directly applying them
+;; to the result of each subsequent application, with the last argument
+;; being passed to the penultimate application as a (single) argument.
+;; This is specialized to our implementation of stream fusion in the
+;; arguments it expects and how it uses them.
+(define-syntax inline-compose1
+  (syntax-rules ()
+    [(_ f) f]
+    [(_ [op f] rest ...) (op f (inline-compose1 rest ...))]))
+
+(begin-for-syntax
```

Comment:
This is a long block: should it perhaps be in a separate module that can be
required `for-syntax`? (I was originally going to suggest using
`define-for-syntax where possible, but you'd still have to have some
`begin-for-syntax` blocks.)